runtime-sdk/modules/evm/src/state.rs

```diff
@@ -0,0 +1,97 @@
+use crate::types::H160;
+
+use oasis_runtime_sdk::{context::Context, storage};
+
+/// Prefix for Ethereum account code in our storage (maps H160 -> Vec<u8>).
+pub const CODES: &[u8] = &[0x01];
+/// Prefix for Ethereum account storage in our storage (maps H160||H256 -> H256).
+pub const STORAGES: &[u8] = &[0x02];
+/// Prefix for Ethereum block hashes (only for last BLOCK_HASH_WINDOW_SIZE blocks
+/// excluding current) storage in our storage (maps Round -> H256).
+pub const BLOCK_HASHES: &[u8] = &[0x03];
+/// Prefix for Ethereum account storage in our confidential storage (maps H160||H256 -> H256).
+pub const CONFIDENTIAL_STORAGES: &[u8] = &[0x04];
+
+/// Confidential store key pair ID domain separation context base.
+pub const CONFIDENTIAL_STORE_KEY_PAIR_ID_CONTEXT_BASE: &[u8] = b"oasis-runtime-sdk/evm: state";
+const CONTEXT_KEY_CONFIDENTIAL_STORE_INSTANCE_COUNT: &str = "evm.ConfidentialStoreCounter";
+
+/// The number of hash blocks that can be obtained from the current blockchain.
+pub const BLOCK_HASH_WINDOW_SIZE: u64 = 256;
+
+pub fn public_storage<'a, C: Context>(
+    ctx: &'a mut C,
+    address: &'a H160,
+) -> storage::TypedStore<impl storage::Store + 'a> {
+    storage::TypedStore::new(storage::HashedStore::<_, blake3::Hasher>::new(
+        contract_storage(ctx.runtime_state(), STORAGES, address),
+    ))
+}
+
+pub fn confidential_storage<'a, C: Context>(
+    ctx: &'a mut C,
+    address: &'a H160,
+) -> storage::TypedStore<Box<dyn storage::Store + 'a>> {
+    let kmgr_client = ctx
+        .key_manager()
+        .expect("key manager must be available to use confidential storage");
+    let key_id = oasis_runtime_sdk::keymanager::get_key_pair_id(&[
+        CONFIDENTIAL_STORE_KEY_PAIR_ID_CONTEXT_BASE,
+        address.as_ref(),
+    ]);
+    let keypair = kmgr_client
+        .get_or_create_keys(key_id)
+        .expect("unable to retrieve confidential storage keys");
```

Comment:
It is even better to propagate a proper aborting error (e.g. see the `Abort` error in the `contracts` module) as a panic will trigger a runtime restart.

Reply:
> Why not modify dispatch_tx to also std::panic::catch_unwind and turn it into a batch failure? That would solve the symptom.

Ah yes, that would also work, can you open an issue for this?